# GDGoC ML Study Jam - Chapter 1 Challenge
## **Topic: The Smartphone Analyst**

Selamat datang, Data Detective! 🕵️‍♂️

Hari ini, kita tidak hanya belajar koding, tapi belajar **memecahkan masalah menggunakan data**.

**Skenario:**
Kamu adalah Lead Data Analyst di sebuah toko gadget besar. Bosmu memberikan dataset spesifikasi smartphone dan memintamu menjawab **10 Pertanyaan Bisnis** untuk strategi penjualan bulan depan.

**Aturan Main:**
1.  **Bebas Berkreasi:** Gunakan metode apapun (Filtering, Sorting, Grouping, atau Visualisasi Grafik).
2.  **Wajib Ada Insight:** Setelah kode dijalankan, tulis kesimpulanmu di cell Teks di bawahnya.
3.  **Have Fun!**

In [ ]:
# ⚠️ JALANKAN CELL INI TERLEBIH DAHULU!
# Script ini akan membuat dataset dummy untuk latihan kita hari ini.

import pandas as pd
import numpy as np
import random

np.random.seed(42)

# Data Dummy
n_data = 150
brands = ['Samsung', 'Apple', 'Xiaomi', 'Oppo', 'Vivo', 'Realme', 'Infinix', 'Asus', 'Google']
rams = [4, 6, 8, 12, 16, 24] # Ada RAM 24GB buat gaming
storages = [64, 128, 256, 512, 1024]

data = {
    'Brand': [random.choice(brands) for _ in range(n_data)],
    'Model': [f"Phone {i}" for i in range(n_data)],
    'RAM_GB': [random.choice(rams) for _ in range(n_data)],
    'Storage_GB': [random.choice(storages) for _ in range(n_data)],
    'Rating': [round(random.uniform(3.0, 5.0), 1) for _ in range(n_data)]
}

df = pd.DataFrame(data)

# Bikin harga agak logis -> (RAM gede + Storage gede = Mahal) + sedikit random noise
df['Price_Juta'] = (df['RAM_GB'] * 0.5) + (df['Storage_GB'] * 0.015) + np.random.randint(1, 4, n_data)
df['Price_Juta'] = df['Price_Juta'].round(1)

# Tambahin Data -> "Aneh" (Anomaly)
df.loc[148] = ['Nokia', 'Nokia 3310 Reborn', 2, 16, 50.0, 5.0] # HP kentang harga sultan
df.loc[149] = ['Unknown', 'Prototype X', 128, 2048, 0.5, 1.0] # Spek dewa harga permen

print("Dataset 'smartphones' berhasil dibuat!")
print(f"Total Data: {len(df)} baris.")
df.head(10)

Dataset 'smartphones' berhasil dibuat!
Total Data: 150 baris.


Brand    Model  RAM_GB  Storage_GB  Rating  Price_Juta
0   Xiaomi  Phone 0       8          64     3.5         8.0
1     Asus  Phone 1      24         512     4.1        20.7
2     Vivo  Phone 2       8          64     3.3         8.0
3  Infinix  Phone 3       8         512     4.7        14.7
4   Google  Phone 4      16         512     3.3        16.7
5   Xiaomi  Phone 5      24         512     4.4        20.7
6   Google  Phone 6      16         256     4.6        14.8
7   Xiaomi  Phone 7       4         512     4.8        11.7
8   Google  Phone 8       4         512     3.4        12.7
9   Xiaomi  Phone 9       6         512     4.9        13.7

---
### **Quest 1: Cek Ombak (Overview)**
Sebelum masuk ke analisis dalam, coba cek dulu isi gudang kita.
1. Ada berapa baris dan kolom di data ini?
2. Apa saja nama kolomnya?
3. Tampilkan statistik dasarnya (Rata-rata, Max, Min).

*Hint: Gunakan `.shape`, `.columns`, dan `.describe()`.*

In [ ]:
# 1. Ada berapa baris dan kolom di data ini?
print("1. Baris dan Kolom")
print(f"Jumlah Baris: {df.shape[0]}")
print(f"Jumlah Kolom: {df.shape[1]}")

# 2. Apa saja nama kolomnya?
print("\n2. Nama Kolom")
print(df.columns.tolist())

# 3. Tampilkan statistik dasarnya (Rata-rata, Max, Min).
print("\n3. Statistik Dasar")
print(df.describe())

1. Baris dan Kolom
Jumlah Baris: 150
Jumlah Kolom: 6

2. Nama Kolom
['Brand', 'Model', 'RAM_GB', 'Storage_GB', 'Rating', 'Price_Juta']

3. Statistik Dasar
           RAM_GB   Storage_GB      Rating  Price_Juta
count  150.000000   150.000000  150.000000  150.000000
mean    12.760000   421.226667    4.348000   14.148000
std     11.791636   355.422032    3.813482    6.265916
min      2.000000    16.000000    0.500000    1.000000
25%      6.000000   128.000000    3.525000    8.925000
50%     10.000000   256.000000    4.000000   13.700000
75%     16.000000   512.000000    4.675000   17.775000
max    128.000000  2048.000000   50.000000   30.400000


Dari hasil df.describe(), terlihat ada anomali pada data. Rata-rata rating terlihat normal, namun nilai Max Rating mencapai 50.0, padahal seharusnya skala rating hanya sampai 5.0. Selain itu, ada smartphone dengan RAM 128 GB yang jauh di atas rata-rata pasar. Ini mengindikasikan adanya data outliers yang perlu dibersihkan nanti.

---
### **Quest 2: Raja Pasar**
Brand apa yang paling banyak memiliki model HP dalam daftar ini?

*Hint: Gunakan `.value_counts()`.*

In [ ]:
# Menghitung jumlah kemunculan setiap Brand
jumlah_per_brand = df['Brand'].value_counts()

print("Peringkat Brand")
print(jumlah_per_brand)

# Paling banyak model hp
model_terbanyak = jumlah_per_brand.idxmax()
jumlahnya = jumlah_per_brand.max()

print(f"\nPaling banyak adalah: {model_terbanyak} dengan total {jumlahnya} model.")

Peringkat Brand
Brand
Xiaomi     22
Vivo       21
Asus       17
Infinix    17
Google     17
Apple      17
Realme     15
Samsung    12
Oppo       10
Nokia       1
Unknown     1
Name: count, dtype: int64

Paling banyak adalah: Xiaomi dengan total 22 model.


Berdasarkan hasil analisis menggunakan .value_counts(), Xiaomi adalah brand yang paling mendominasi dataset dengan total 22 model. Sementara itu, brand Nokia memiliki jumlah model paling sedikit.

---
### **Quest 3: Storage Wars**
Zaman sekarang aplikasi makin besar.
Coba cari HP yang punya **Storage paling besar** (Max Storage). Ada berapa HP yang punya storage segitu?

*Hint: Cari nilai max dulu, lalu filter datanya.*

In [ ]:
# 1. Storage paling besar (Max)
max_storage = df['Storage_GB'].max()
print(f"Kapasitas Storage Terbesar: {max_storage} GB")

# 2. Filter data: HP yang punya storage sebesar max_storage
hp_storage_besar = df[df['Storage_GB'] == max_storage]

# 3. Tampilkan hasil
print(f"Jumlah HP dengan storage besar: {len(hp_storage_besar)} unit")
print("\nDetail HP:")
hp_storage_besar # Menampilkan tabel baris yang lolos filter

Kapasitas Storage Terbesar: 2048 GB
Jumlah HP dengan storage besar: 1 unit

Detail HP:


Brand        Model  RAM_GB  Storage_GB  Rating  Price_Juta
149  Unknown  Prototype X     128        2048     0.5         1.0

Ditemukan bahwa kapasitas storage terbesar dalam dataset adalah 2048 GB (2 TB). Hanya ada 1 unit HP yang memiliki storage sebesar ini, yaitu Brand Unknown dengan model Prototype X. Ini merupakan data anomali karena jauh di atas standar pasar HP saat ini, dan brand-nya pun tidak dikenal.

---
### **Quest 4: Budget vs Premium**
Bos ingin tahu rentang harga pasar.
1. Berapa harga HP **termurah**?
2. Berapa harga HP **termahal**?

*Hint: Gunakan `.min()` dan `.max()` pada kolom harga.*

In [ ]:
# Cari harga termurah dan termahal
harga_min = df['Price_Juta'].min()
harga_max = df['Price_Juta'].max()

print(f"Harga Termurah: {harga_min} Juta")
print(f"Harga Termahal: {harga_max} Juta")

print("\nDetail HP Termurah")
print(df[df['Price_Juta'] == harga_min][['Brand', 'Model', 'RAM_GB', 'Price_Juta']])

print("\nDetail HP Termahal")
print(df[df['Price_Juta'] == harga_max][['Brand', 'Model', 'RAM_GB', 'Price_Juta']])

Harga Termurah: 1.0 Juta
Harga Termahal: 30.4 Juta

Detail HP Termurah
       Brand        Model  RAM_GB  Price_Juta
149  Unknown  Prototype X     128         1.0

Detail HP Termahal
   Brand     Model  RAM_GB  Price_Juta
31  Asus  Phone 31      24        30.4
55  Asus  Phone 55      24        30.4


Harga termurah adalah 1.0 Juta, namun ini dimiliki oleh 'Prototype X' yang memiliki RAM 128GB. Sedangkan harga termahal mencapai sekitar 30-an Juta.

---
### **Quest 5: The Gaming Beast**
Seorang Gamer datang mencari HP. Syaratnya mutlak:
1. **RAM wajib di atas 12 GB.**
2. **Storage minimal 512 GB.**

Tampilkan daftar HP yang memenuhi syarat tersebut!

*Hint: Gunakan Filtering dengan logika AND (`&`).*

In [ ]:
hp_gaming = df[(df['RAM_GB'] > 12) & (df['Storage_GB'] >= 512)]

print(f"Ada {len(hp_gaming)} HP Gaming")
print("\nDaftar HP Gaming")
hp_gaming

Ada 26 HP Gaming

Daftar HP Gaming


Brand        Model  RAM_GB  Storage_GB  Rating  Price_Juta
1       Asus      Phone 1      24         512     4.1        20.7
4     Google      Phone 4      16         512     3.3        16.7
5     Xiaomi      Phone 5      24         512     4.4        20.7
16   Infinix     Phone 16      24        1024     4.2        29.4
31      Asus     Phone 31      24        1024     3.5        30.4
34   Infinix     Phone 34      24         512     4.6        22.7
40      Asus     Phone 40      16         512     4.2        16.7
46    Realme     Phone 46      16         512     4.8        17.7
55      Asus     Phone 55      24        1024     3.9        30.4
56      Asus     Phone 56      24         512     3.0        22.7
57    Xiaomi     Phone 57      16        1024     4.1        24.4
58     Apple     Phone 58      16         512     4.1        18.7
76      Vivo     Phone 76      16         512     4.8        17.7
80      Vivo     Phone 80      24         512     3.8        22.7
86   Samsung     Phone 86      24         512     3.7        22.7
88      Asus     Phone 88      24        1024     4.8        29.4
90      Asus     Phone 90      16         512     3.7        16.7
94   Samsung     Phone 94      16         512     3.1        16.7
98    Realme     Phone 98      16         512     5.0        16.7
117     Asus    Phone 117      16        1024     4.9        26.4
119     Oppo    Phone 119      24        1024     4.2        29.4
127     Vivo    Phone 127      16         512     4.9        16.7
141     Asus    Phone 141      24         512     4.5        22.7
142   Xiaomi    Phone 142      24        1024     4.3        29.4
146   Xiaomi    Phone 146      16         512     4.0        18.7
149  Unknown  Prototype X     128        2048     0.5         1.0

Dari hasil filtering, ada 26 model HP dengan spesifikasi gaming (RAM 16GB ke atas dan Storage besar) dari brand yang berbeda-beda.

---
### **Quest 6: Sobat Mahasiswa**
Mahasiswa biasanya cari HP yang "Worth it".
Cari HP yang harganya **di bawah 7 Juta**, TAPI **Rating-nya di atas 4.5**.

Ada nggak barangnya?

In [ ]:
# Filter data: Harga < 7 Juta DAN Rating > 4.5
hp_mahasiswa = df[(df['Price_Juta'] < 7) & (df['Rating'] > 4.5)]

print(f"Ditemukan {len(hp_mahasiswa)} HP yang cocok untuk mahasiswa")
print("\nDaftar HP Worth It")
hp_mahasiswa

Ditemukan 7 HP yang cocok untuk mahasiswa

Daftar HP Worth It


Brand              Model  RAM_GB  Storage_GB  Rating  Price_Juta
47      Vivo           Phone 47       8         128     4.6         6.9
79   Infinix           Phone 79       6         128     4.6         6.9
95     Apple           Phone 95       4          64     4.7         4.0
99    Google           Phone 99       8         128     4.6         6.9
131  Infinix          Phone 131       8          64     4.7         6.0
143  Infinix          Phone 143       4          64     4.8         5.0
148    Nokia  Nokia 3310 Reborn       2          16    50.0         5.0

Setelah difilter, ada 7 model HP yang cocok untuk mahasiswa dari 5 brand hp yang berbeda

---
### **Quest 7: Fanboy Apple**
Ada pelanggan yang cuma mau beli **Apple**.
Tampilkan semua HP merk Apple, lalu urutkan dari yang **paling mahal ke paling murah**.

*Hint: Filter `Brand == 'Apple'`, lalu `sort_values`.*

In [ ]:
# 1. Filter: Brand Apple
apple_only = df[df['Brand'] == 'Apple']

# 2. Sort: Urutkan Besar ke Kecil
apple_sorted = apple_only.sort_values(by='Price_Juta', ascending=False)

print(f"Ditemukan {len(apple_sorted)} model Apple.")
print("\nKatalog Apple (Termahal ke Termurah)")
apple_sorted

Ditemukan 17 model Apple.

Katalog Apple (Termahal ke Termurah)


Brand      Model  RAM_GB  Storage_GB  Rating  Price_Juta
102  Apple  Phone 102       8        1024     4.2        22.4
22   Apple   Phone 22       6        1024     4.9        20.4
26   Apple   Phone 26       6        1024     4.0        19.4
58   Apple   Phone 58      16         512     4.1        18.7
11   Apple   Phone 11      24          64     4.6        16.0
125  Apple  Phone 125      24          64     3.2        14.0
107  Apple  Phone 107      16          64     4.2        12.0
73   Apple   Phone 73       4         512     4.9        11.7
145  Apple  Phone 145      12         128     3.4        10.9
89   Apple   Phone 89       4         512     4.2        10.7
38   Apple   Phone 38       4         512     4.8        10.7
140  Apple  Phone 140      12          64     4.3        10.0
33   Apple   Phone 33       6         256     4.3         8.8
15   Apple   Phone 15       8         256     4.9         8.8
93   Apple   Phone 93       6          64     4.0         7.0
109  Apple  Phone 109       4         128     3.3         5.9
95   Apple   Phone 95       4          64     4.7         4.0

Berdasarkan data yang difilter, varian HP Apple termahal di toko harganya 22.4 Juta. Sedangkan varian termurahnya ada di angka 4.0 Juta. Dengan mengurutkan seperti ini, kita bisa langsung menawarkan opsi terbaik saat pelanggan Apple Fanboy datang menanyakan model terbaik.

---
### **Quest 8: Rata-Rata Harga per Brand (Level Medium)**
Brand mana yang rata-rata harga produknya paling mahal?
Coba hitung rata-rata harga (`Price_Juta`) untuk setiap `Brand`.

*Hint: Ini butuh Grouping. Gunakan `.groupby('Brand')['Price_Juta'].mean()`.*

In [ ]:
# 1. Grouping
rata_rata_harga = df.groupby('Brand')['Price_Juta'].mean()

# 2. Sorting: Urutkan dari rata-rata termahal ke termurah
rata_rata_sorted = rata_rata_harga.sort_values(ascending=False)

print("Rata-Rata Harga HP per Brand (Juta)")
print(rata_rata_sorted)

print(f"\nBrand dengan harga rata rata HP tertinggi adalah {rata_rata_sorted.idxmax()}")

Rata-Rata Harga HP per Brand (Juta)
Brand
Asus       19.029412
Google     15.923529
Samsung    15.266667
Xiaomi     14.559091
Oppo       13.880000
Realme     13.000000
Infinix    12.576471
Apple      12.435294
Vivo       12.357143
Nokia       5.000000
Unknown     1.000000
Name: Price_Juta, dtype: float64

Brand dengan harga rata rata HP tertinggi adalah Asus


Berdasarkan rata-rata harga, brand Asus menempati posisi pertama sebagai brand paling mahal/premium, dengan rata-rata harga sekitar 19 Juta. Hal ini menunjukkan bahwa stok Asus didominasi oleh varian spesifikasi tinggi. Sebaliknya, brand Nokia dan Unknown memiliki rata-rata harga terendah, cocok untuk target pasar entry-level.

---
### **Quest 9: Value for Money (Feature Engineering)**
Kita mau cari HP yang "Spek Dewa Harga Kaki Lima".
1. Buat kolom baru bernama `Skor_Performa` = `RAM_GB` + `Storage_GB`.
2. Buat kolom baru bernama `Value_Skor` = `Skor_Performa` / `Price_Juta`.
3. Tampilkan 5 HP dengan `Value_Skor` tertinggi.

*Hint: Kamu bisa melakukan operasi matematika antar kolom secara langsung.*

In [ ]:
# 1. Buat kolom 'Skor_Performa' (RAM + Storage)
df['Skor_Performa'] = df['RAM_GB'] + df['Storage_GB']

# 2. Buat kolom 'Value_Skor' (Performa banding Harga)
df['Value_Skor'] = df['Skor_Performa'] / df['Price_Juta']

# 3. Urutkan dan Tampilkan 5 terbaik
top_value = df.sort_values(by='Value_Skor', ascending=False).head(5)

print("Top 5 HP dengan value terbaik")
print(top_value[['Brand', 'Model', 'RAM_GB', 'Storage_GB', 'Price_Juta', 'Value_Skor']])

Top 5 HP dengan value terbaik
       Brand        Model  RAM_GB  Storage_GB  Price_Juta   Value_Skor
149  Unknown  Prototype X     128        2048         1.0  2176.000000
26     Apple     Phone 26       6        1024        19.4    53.092784
18    Google     Phone 18       4        1024        19.4    52.989691
60    Google     Phone 60       4        1024        19.4    52.989691
103     Oppo    Phone 103       8        1024        20.4    50.588235


Setelah menghitung Value_Skor, peringkat 1 ditempati oleh 'Prototype X'. Ini terjadi karena dia memiliki Storage sangat besar (2048 GB) tapi harganya sangat murah. Skornya menjadi sangat tinggi secara tidak wajar. Di tempat ke 2 ada brand HP dari Apple yang menawarkan storage besar dengan harga kompetitif.

---
### **Quest 10: The Detective (Final Boss)**
Ada laporan data error di sistem.
1. Cari HP yang harganya **di atas 40 Juta** (Kemahalan/Mustahil).
2. Cari HP yang harganya **di bawah atau sama dengan 1 Juta** tapi RAM-nya **di atas 100 GB** (Ngawur).

Temukan data penyusup ini!

*Hint: Gunakan filtering untuk mencari outlier.*

In [ ]:
# 1: Harga di atas 40 Juta
error_1 = df['Price_Juta'] > 40

# 2: Harga <= 1 Juta dan RAM > 100 GB
error_2 = (df['Price_Juta'] <= 1) & (df['RAM_GB'] > 100)

data_error = df[error_1 | error_2]

print("LAPORAN ERROR DITEMUKAN")
print(f"Jumlah Data Error: {len(data_error)}")
data_error

LAPORAN ERROR DITEMUKAN
Jumlah Data Error: 1


Brand        Model  RAM_GB  Storage_GB  Rating  Price_Juta  \
149  Unknown  Prototype X     128        2048     0.5         1.0   

     Skor_Performa  Value_Skor  
149           2176      2176.0

# 🎉 Mission Accomplished!

Selamat! Kamu sudah menyelesaikan 10 Misi.
Jangan lupa:
1.  **Kumpulkan di Google Classroom**
2.  Kumpulkan dalam bentuk link colab (Can Edit Access).

Sampai jumpa di Chapter 2!